README.md
# ranking-players


ChangeLog.md
# Changelog for ranking-players

## Unreleased changes


test/Spec.hs
import           Test.Hspec
import           Internal.ExpectationNet


--data NodeDescription = GF String Double Double | DGF String String Double | GrT String (String,String) Double

nodes =
        [ GF "Fred_Skill" 100 5
        , GF "Jill_Skill" 120 40
        , DGF "Fred_perf" "Fred_Skill" 5
        , DGF "Jill_perf" "Jill_Skill" 5
        , GrT "Jill_Win" ("Fred_perf", "Jill_perf") 0.5
        ]

graph = generateGraph nodes 

testGenerateGraph :: IO ()
testGenerateGraph = do
  let expectedGraph =
        [ GFNode 0 "Fred_Skill" (100.0, 5.0) [1]
        , GNode 1 "Fred_Skill" (100.0, 5.0) [4, 0]
        , GFNode 2 "Jill_Skill" (120.0, 40.0) [3]
        , GNode 3 "Jill_Skill" (120.0, 40.0) [6, 2]
        , DGFNode 4 "Fred_perf" "Fred_Skill" (0.0, 5.0) [1, 5]
        , GNode 5 "Fred_perf" (0.0, 0.0) [8, 4]
        , DGFNode 6 "Jill_perf" "Jill_Skill" (0.0, 5.0) [3, 7]
        , GNode 7 "Jill_perf" (0.0, 0.0) [8, 6]
        , GTNode 8 "Jill_Win" (5, 7) [9]
        , VNode 9 ("Jill_Win", 0.5) [8]
        ]
  print graph
  print $ setObservation "Jill_Win" False graph
  shouldBe graph expectedGraph

testDistFunctions :: IO ()
testDistFunctions = do
      let xs = [-30,-29.99..300]
      let gauss = norm $ zip xs (map (normal ("test", 100, 5)) xs)
      let mu = mean gauss
      print $ "mu " ++ show mu
      print $ "stdev " ++ show ( stdev gauss mu)

testDivisionGaussian :: IO ()
testDivisionGaussian = do
      let f_skill = ("Fskill", 110, 40)
      let j_skill = ("Jskill", 120, 5) 
      let norm1 = normal f_skill
      let norm2 = normal j_skill
      let xs = [-1000,-999.9..1000]
      let normal1 = zip xs $ map norm1 xs
      let normal2 = zip xs $ map norm2 xs
      let divide =  norm $ map (\((a,b),(c,d)) -> if(d == 0) then (a,0) else (a,b/d) ) $ zip normal2 normal1
      --print divide
      let mu = mean divide
      print $ "mu " ++ show mu
      print $ "stdev " ++ show (stdev divide mu)
      print $ normalDivision j_skill f_skill


testApproxGaussian :: IO ()
testApproxGaussian = do
      let f_skill = ("Fskill", 100, 5)
      let j_skill = ("Jskill", 120, 40)
      let cum = cumulative f_skill True
      let cum2 = cumulative j_skill True 
      print $ approxGaussian j_skill cum
      print $ approxGaussian f_skill cum2
      

testGetMessage :: IO ()
testGetMessage = do
  let n_graph = setObservation "Jill_Win" True graph
  print "After Jill wins"
  print $ getMessage (0) "Fred_skill" (n_graph !! 1) n_graph
  print $ getMessage (4) "Fred_skill" (n_graph !! 1) n_graph
  print $ getMessage (-1) "Fred_skill" (n_graph !! 1) n_graph
  print $ getMessage (-1) "Jill_skill" (n_graph !! 3) n_graph
  print $ getMessage 5 "Fred_perf" (n_graph !! 8) n_graph
  print $ getMessage 7 "Jill_perf" (n_graph !! 8) n_graph
  print "After Jill looses"
  let l_graph = setObservation "Jill_Win" False graph
  print $ getMessage (-1) "Fred_skill" (l_graph !! 1) l_graph
  print $ getMessage (-1) "Jill_skill" (l_graph !! 3) l_graph

-- evalVariable variable observations graph 
testEvalVariable :: IO ()
testEvalVariable = do
      print "Jill Wins:"
      print $ evalVariable "Fred_Skill" ["Jill_Win"] [True] graph
      print $ evalVariable "Jill_Skill" ["Jill_Win"] [True] graph
      print "Jill looses:"
      print $ evalVariable "Fred_Skill" ["Jill_Win"] [False] graph
      print $ evalVariable "Jill_Skill" ["Jill_Win"] [False] graph

testEvalVariables :: IO ()
testEvalVariables = do
      let tenWins = replicate 10 True
      let tenLooses = replicate 10 False
      print "Jill Wins 10:"
      print $ snd $ evalVariables ["Fred_Skill","Jill_Skill"] ["Jill_Win"] [tenWins] graph
      print "Jill looses 10:" 
      print $ snd $ evalVariables ["Fred_Skill","Jill_Skill"] ["Jill_Win"] [tenLooses] graph

main :: IO ()
main = do
  testGenerateGraph
  testDistFunctions
  testDivisionGaussian
  testApproxGaussian
  putStrLn "\nTest Get message\n"
  testGetMessage
  putStrLn "\nEvalVariable\n"
  testEvalVariable
  putStrLn "\nEvalVariabls\n"
  testEvalVariables


src/Internal/FakeDataBase.hs
{-# LANGUAGE DeriveGeneric, OverloadedStrings, OverloadedLabels, TypeOperators #-}
module Internal.FakeDataBase where
import           Database.Selda
import           Database.Selda.SQLite
import Database.Selda.Unsafe
import System.Random
import Control.Monad
import Data.Random
import           Data.Random.Distribution.Normal
import           Data.Random.RVar
import           Data.Random.Source.DevRandom

sampleNormal :: Int -> Double -> Double -> IO [Double]
sampleNormal nr mean std = replicateM nr (runRVar (Data.Random.normal mean std) DevRandom :: IO Double)

playerStrengths :: IO [Double]
playerStrengths = sampleNormal (length names) 100 15--replicateM (length names) (randomRIO (0::Double, 1))

results :: IO [(Int,Int,Int)]
results = do
  strengths <- playerStrengths
  print $ "True Skill values :"
  mapM_ print $ zip [1..] strengths
  replicateM 10000 (generateResult strengths)

generateResult :: [Double] -> IO (Int,Int,Int)
generateResult strengths = do
  let ids = [1..(length names +1)]
  pID1 <- randomRIO (1::Int, length names)
  let (xs,y:ys) = splitAt (pID1-1) ids
  pID2 <- (\nr ->  (xs ++ ys) !! nr) <$> randomRIO (0::Int, length names -2 )  
  let str1 = strengths !! (pID1-1)
  let str2 = strengths !! (pID2-1)
  perf1 <- head <$> sampleNormal 1 0 5--randomRIO(0::Double,1)
  perf2 <- head <$> sampleNormal 1 0 5--randomRIO(0::Double,1)
  return $ 
    if str1 + perf1 > str2 + perf2 
      then (pID1, pID2, pID1)
      else (pID1, pID2, pID2)

players :: Table (RowID :*: Text :*: Double :*: Double)
players = table "players" $ autoPrimary "id" :*: required "name" :*: required "mu" :*: required "std"

games :: Table (RowID :*: RowID :*: RowID :*: RowID)
games = table "games" $ autoPrimary "id" :*: required "player1" :*: required "player2" :*: required "winner"

printPlayer :: Int -> IO ()
printPlayer id = withSQLite "my_database.sqlite" $ do
  plz <- query $ do
    (id':*: name :*:mu :*:std) <- select players
    restrict(id' .== literal ( unsafeRowId id))
    return (id':*: name :*:mu :*:std)
  liftIO (print plz)


numberGames :: IO Int
numberGames = withSQLite "my_database.sqlite" $ do
  plz <- query $ aggregate $ do
    (id :*: _) <- select games
    return $ count id
  games' <- query $ select games
  liftIO (return $ head plz)

getGameRecord :: Int -> IO (RowID,RowID,RowID,Double,Double,Double,Double) 
getGameRecord nr = withSQLite "my_database.sqlite" $ do
  values <- query $ getGameRecord' nr
  let (player1 :*: player2 :*: winner :*: mu :*: std :*: mu2 :*: std2) = head values
  liftIO (return  (player1,player2,winner,mu,std,mu2,std2))  


getGameRecord' nr = do
  (gid :*: player1 :*: player2 :*: winner) <- select games
  (id1 :*: _ :*: mu :*: std) <- select players
  restrict ( player1  .== id1 )
  (id2 :*: _ :*: mu2 :*: std2) <- select players
  restrict ( player2 .== id2 )
  restrict (gid .== literal (unsafeRowId nr))
  return (player1 :*: player2 :*: winner :*: mu :*: std :*: mu2 :*: std2)

resetPlayers :: IO ()
resetPlayers = withSQLite "my_database.sqlite" $ do
  resetPlayers'

resetPlayers' :: SeldaM ()
resetPlayers' = do
  update_ players (const true)
                  (\(id :*: name :*: _ :*: _ ) -> id :*: name :*: float 100.0 :*: float 50.0)

updatePlayer :: RowID -> Double -> Double -> IO()
updatePlayer id mu std = withSQLite "my_database.sqlite" $ do
  updatePlayer' id mu std

updatePlayer' :: RowID -> Double -> Double -> SeldaM ()
updatePlayer' id mu std = do
  update_ players (\(id' :*: _ :*: _) -> literal id .== id')
                  (\(id :*: name :*: _ :*: _ ) -> id :*: name :*: float mu :*: float std)

setup :: SeldaM ()
setup = do
  createTable players
  createTable games
  
initDatabase :: IO ()
initDatabase = withSQLite "my_database.sqlite" $ do
    teardown
    setup
    res <- liftIO results
    insertGames res
    insertPlayers

insertPlayers :: SeldaM ()
insertPlayers = do
  insert_ players $ map (\name -> def :*: name :*: (100::Double) :*: (15::Double)) names
  
insertGames :: [(Int,Int,Int)] -> SeldaM ()
insertGames results = do
  insert_ games $ map(\(a,b,c) -> def :*:  unsafeRowId a :*: unsafeRowId b :*: unsafeRowId c) results 

teardown :: SeldaM ()
teardown = do
  tryDropTable games
  tryDropTable players

names :: [Text]
names =
  [ "Elias"
  , "Liam"
  , "Linus"
  , "Julia"
  , "Juergen"
  , "Jonas"
  , "CrazyCat"
  , "SleepyLion"
  , "Thomas"
  , "Pauls"
  , "Johannes"
  , "Marius"
  , "Tim"
  , "Maya"
  , "Simone"
  , "Andrea"
  , "Lukas"
  , "Verona"
  , "Kroete"
  , "KillerRabe"
  , "TheFrog"
  , "Grenui"
  , "Mirama"
  , "Miriam"
  , "Daniela"
  , "Natascha"
  , "Chantal"
  , "Tobias"
  , "Sebastian"
  , "Arnold"
  , "Mika"
  , "TheShadow"
  , "LuckyLuke"
  , "GraberPaba"
  , "HerrVonHoellestein"
  , "Winnie"
  , "Ursula"
  , "Gerald"
  , "Ulrike"
  , "PlayerBoy"
  , "Volker"
  , "Anna"
  , "Carla"
  , "James"
  , "Bond"
  , "Yogi"
  , "Napoleon"
  , "Riddler"
  , "Bierbauch"
  , "Butter"
  , "MisterX"
  , "Holger"
  , "Erich"
  , "Baldrian"
  , "Baldur"
  , "Jayn"
  , "Frank"
  , "Walter"
  , "Christiane"
  , "Matthias"
  , "Mina"
  , "Roscharch"
  , "Mika"
  , "Joe"
  , "Stepfhen"
  , "Dogbert"
  , "Dagobert"
  , "Hodor"
  , "Sam"
  , "Stefanie"
  , "TwitterGirl"
  , "Legolas"
  , "Kalle"
  , "LatteIgel"
  , "Barbara"
  , "Verona"
  , "Wilma"
  , "Sanders"
  , "Finn"
  , "Leopold"
  , "Bernie"
  , "Fritz"
  , "Peter"
  , "Harold"
  , "Sansa"
  , "Arya"
  , "Snow"
  , "Tyrion"
  , "Catelyn"
  , "Eddard"
  , "Robb"
  , "Bran"
  , "Tywin"
  , "Baratheon"
  , "Theon"
  , "Renly"
  , "Stannis"
  , "Tyenne"
  , "GreenEyes"
  , "FinalChapter"
  ]

src/Internal/ExpectationNet.hs
module Internal.ExpectationNet where

import           Data.List
import           System.Random
import Data.Number.Erf

type Bern = (String, Double)

--type Factor = ([String], [[Bool]], [Double])

data Node = GNode Int String (Double,Double) [Int]  | GFNode Int String (Double,Double) [Int]
            | DGFNode Int String String (Double,Double) [Int]   | GTNode Int String (Int,Int) [Int] | VNode Int Bern [Int]  | ONode Int Bern [Int] deriving (Show, Eq)


data NodeDescription = GF String Double Double | DGF String String Double | GrT String (String,String) Double


type Graph = [Node]

generateGraph :: [NodeDescription] -> Graph
generateGraph descriptions =
  foldl (\graph des -> addNodePair graph des) [] descriptions

addNodePair :: Graph -> NodeDescription -> Graph
addNodePair graph (GF name mu std) =
  graph
    ++ [ GFNode nId name (mu, std) [nId + 1]
       , GNode (nId + 1) name (mu, std) [nId]
       ]
  where nId = length graph
addNodePair graph (DGF name1 name2 std) =
  n_graph
    ++ [ DGFNode nId name1 name2 (0, std) [id, nId + 1]
       , GNode (nId + 1) name1 (0, 0) [nId]
       ]
 where
  nId     = length graph
  id      = findId name2 graph
  n_graph = updateGraph nId [id] graph
addNodePair graph (GrT name (lesser, larger) value) =
  n_graph
    ++ [ GTNode nId name (n_lesser, n_larger) [nId + 1]
       , VNode (nId + 1) (name, value) [nId]
       ]
 where
  nId      = length graph
  n_lesser = findId lesser graph
  n_larger = findId larger graph
  n_graph  = updateGraph nId [n_lesser, n_larger] graph

findId :: String -> Graph -> Int
findId name graph = head $ concatMap (nodeToID name) graph

nodeToID :: String -> Node -> [Int]
nodeToID searched (GNode id name _ _) = if name == searched then [id] else []
nodeToID _        _                   = []

updateGraph :: Int -> [Int] -> Graph -> Graph
updateGraph id ids graph = foldl updateGraph' graph ids
  where updateGraph' graph to = map (addId id to) graph


addId :: Int -> Int -> Node -> Node
addId from to node@(GNode id n p ids) =
  if id == to then (GNode id n p (from : ids)) else node
addId from to node = node

type NormDist = (String,Double,Double)


normalProduct :: NormDist -> NormDist -> NormDist
normalProduct (name,mu1,std1) (__,mu2,std2)  = (name, mu, sqrt var)
  where 
    mu = (mu1 * std2^2 + mu2 * std1^2) / (std1^2 + std2^2)
    var = (std1*std2)^2 / (std1^2+std2^2)

normalDivision :: NormDist -> NormDist -> NormDist
normalDivision (name,mu1,std1) (__,mu2,std2)  = (name, mu, sqrt var)
  where 
    mu = (mu1 * std2^2 - mu2 * std1^2) / (std2^2 - std1^2)
    var = (std1*std2)^2 / (std2^2-std1^2)


convolution :: NormDist -> NormDist -> NormDist
convolution (name,mu1,std1) (__,mu2,std2) = (name, mu1+mu2, std1+std2)  

cumulative :: NormDist -> Bool -> (Double -> Double)
cumulative (_, mu, std) wins  = if wins then f else (1-) . f 
  where 
    f x = 1/2*(1+ erf( (x-mu)/(sqrt (2* std^2))))

normal :: NormDist -> (Double -> Double)
normal (_,mu,std) = (\x -> 1/ (sqrt ( 2* pi * std^2 )) * exp( - (x-mu)^2/(2*std^2)) )

norm :: [(Double,Double)] -> [(Double,Double)]
norm xs = map(\(a,b) -> (a,b/area)) xs
  where
    pairs = zip xs (tail xs)
    area = sum $ map(\((a,b),(c,d)) -> (c-a) * (b+d)/2 ) pairs
  
mean :: [(Double,Double)] -> Double
mean dist = mu
  where
    pairs = zip dist (tail dist)
    mu = sum $ map(\((a,b),(c,d)) -> (c-a) * (b+d)/2 * (a+c)/2 ) pairs

stdev :: [(Double,Double)] -> Double -> Double
stdev dist mu = sqrt $ eX2 - mu^2 
  where 
    pairs = zip dist (tail dist)
    eX2 = sum $ map(\((a,b),(c,d)) -> (c-a) * (b+d)/2 * ((a+c)/2)^2 ) pairs

approxGaussian :: NormDist -> (Double -> Double) -> NormDist
approxGaussian dist@(_,mu',std') cumul = normalDivision ("approx",mu, dev) dist
  where
    start = mu' - 4 * std'
    end = mu' + 4 *std'
    --xs = [start..end]
    xs = [start,(start+std'/100)..end]
    cum = map cumul xs
    normDist = map (normal dist) xs
    product = map(uncurry (*) ) $ zip cum normDist
    normProduct = norm $ zip xs product
    mu = mean normProduct
    dev = stdev normProduct mu
    

getMessage :: Int -> String -> Node -> Graph -> NormDist
getMessage sender name1 (GNode id name2 (mu,std) ids) graph = foldr1 normalProduct norms
  where 
    froms = filter (/=sender) ids 
    norms = map (\x -> getMessage id name2 (graph !! x) graph) froms   
getMessage sender name (GFNode _ _ (mu,std) _) graph = (name,mu,std)
getMessage sender name (DGFNode id own other (_,std) ids) graph = 
  convolution (name,0,std) dist 
  where
      from = head $ filter (/=sender) ids
      dist = getMessage id own (graph !! from) graph
      sdist = getMessage id other (graph !! sender) graph
getMessage sender name1 (GTNode id name2 (lesser,larger) ids) graph = (name1,mu,std)
  where
    (truthValue,_,_) = getMessage id name2 (graph !! (id+1)) graph
    isGreater = if truthValue == "true" then  True else False
    lessDist = getMessage id name2 (graph !! lesser) graph
    larDist = getMessage id name2 (graph !! larger) graph
    cumul = if sender == larger 
      then
        cumulative lessDist isGreater
      else
        cumulative larDist (not isGreater)  
    (_,mu,std) = if sender == larger
      then
        approxGaussian larDist cumul
      else
        approxGaussian lessDist cumul
 
getMessage sender _ (ONode id (_,value) ids) graph = 
  if value > 0.5
    then ("true", 0,0)
    else ("false" ,0,0)

getMessage sender _ (VNode id (_,value) ids) graph = 
  if value > 0.5
    then ("true",0,0)
    else ("false",0,0)

evalVariables :: [String] -> [String] -> [[Bool]] -> Graph -> (Graph,[NormDist])
evalVariables variables observations [[]] graph = (graph,dists)
  where 
    dists = map (readDistFromGraph graph) variables
evalVariables variables observations values graph = 
  evalVariables variables observations restValues upGraph 
  where
    actValues = map head values
    evals = map (\name -> evalVariable name observations actValues graph) variables
    restValues = map tail values
    upGraph = replaceVariables evals graph

readDistFromGraph :: Graph -> String-> NormDist
readDistFromGraph graph name = (name,mu,std)
  where
    id = findId name graph
    GFNode _ _ (mu,std) _  = graph !! (id-1)
  
replaceVariables :: [NormDist] -> Graph -> Graph
replaceVariables dists graph = foldr replaceVariable graph dists

replaceVariable :: NormDist -> Graph -> Graph
replaceVariable (name,mu,std) graph = before ++ (n_node:after)
  where
    id = findId name graph
    (before,(GFNode _ _ _ ids):after ) = splitAt (id-1) graph
    n_node = GFNode (id-1) name (mu,std) ids 

evalVariable :: String -> [String] -> [Bool] -> Graph -> NormDist
evalVariable variable observations values graph = getMessage (-1) variable (obsGraph !! id) obsGraph  
  where 
    obsGraph = foldr (\(obs,value) g -> setObservation obs value g) graph $ zip observations values 
    id = findId variable obsGraph
    

toId :: String -> Node -> [(Int,[Int])]
toId name1 node@(VNode id (name,_) ids ) = 
  if name == name1
    then [(id,ids)]
    else [] 
toId name1 node@(ONode id (name,_) ids ) = 
  if name == name1
    then [(id,ids)]
    else [] 
toId _ _ = []

findOVNodeId :: String -> Graph -> (Int,[Int])
findOVNodeId name graph = head $ concatMap (toId name) graph



setObservation :: String -> Bool -> Graph -> Graph
setObservation name obs graph = before ++ (n_node:xs)
  where
    (id,ids) = findOVNodeId name graph
    (before,found:xs) = splitAt id graph
    value = if obs then 1 else 0 
    n_node = ONode id (name,value) ids

app/Main.hs
module Main where

import Data.Random.Distribution.Normal
import Data.Random.RVar
import Data.Random
import Data.Random.Source.DevRandom
import Control.Monad
import Graphics.Rendering.Chart.Backend.Cairo
import Graphics.Rendering.Chart.Easy
import Graphics.Rendering.Chart.Plot.Lines
import Data.Number.Erf
import GHC.Float
import Data.List
import Internal.ExpectationNet
import Control.Monad.IO.Class
import Internal.FakeDataBase

sampleStdNormal :: Int -> IO [Double]
sampleStdNormal nr = replicateM nr (runRVar stdNormal DevRandom :: IO Double)

percentageInDeviation :: Double -> [Double] -> Double
percentageInDeviation deviation sample = fromIntegral inDeviation
  / fromIntegral (length sample)
  where inDeviation = length $ filter (\x -> abs x < deviation) sample


task1 :: IO ()
task1 = do
  sample <- sampleStdNormal 1000000
  print $ "In 1 std " ++ ( show $ percentageInDeviation 1 sample )
  print $ "In 2 std " ++ ( show $ percentageInDeviation 2 sample )
  print $ "In 3 std " ++ ( show $ percentageInDeviation 3 sample )

hist :: [Double] -> PlotHist Double Double
hist sample = defaultNormedPlotHist
  { _plot_hist_title     = ""
  , _plot_hist_values    = sample
  , _plot_hist_bins      = 500
  }

stdNormalPdf :: [Double] -> [(Double, Double)]
stdNormalPdf xs  =
  [ (x, 1/ (sqrt (2*pi)) * exp (-x*x/2)) | x <- xs ]

chart sample = toFile def "../visualisation/hist_vs_beta.png" $ do
  layout_title .= "Standard-Normalverteilung"
  plot $ fmap histToPlot $ liftCState $ return $ hist sample 
  plot ( line "" [stdNormalPdf [-4,-3.999..4]])

task2 :: IO ()
task2 = do
    sample <- sampleStdNormal 100000
    chart sample

task3 :: IO ()
task3 = do
    sample <- sampleStdNormal 100000
    let nr = fromIntegral $ length sample
    let mean = sum sample / nr
    let var =  sum (map (\x -> (x-mean)^2) sample ) / nr
    let std =  sqrt var
    print $ "mean" ++ " " ++ show mean
    print $ "var" ++ " " ++ show var
    print $ "std" ++ " " ++ show std

--sampleNormal :: Int -> Double -> Double -> IO [Double]
--sampleNormal nr mean std = replicateM nr (runRVar (Data.Random.normal mean std) DevRandom :: IO Double)


skillPlot :: String -> Double -> [(Double,Double)] -> IO ()
skillPlot title var values = toFile def ("../visualisation/Jill_vs_Fred_std_" ++ show var) $ do
  layout_title .= title
  plot (points "" values)
  plot (line "" [[(-30,-30),(50,50) ]])
  
winRatio :: [(Double,Double)] -> String
winRatio sample = "Jill " ++ show winRatio ++ " vs " ++ "Fred " ++ show (1 - winRatio)
  where
    wins = fromIntegral $ length $ filter (uncurry (<)) sample
    winRatio =  wins / (fromIntegral $ length sample)

sampleGames :: Int -> Double-> IO [(Double,Double)]
sampleGames nr var = do
  sample_jill <- sampleNormal nr 15 var
  sample_fred <- sampleNormal nr 12.5 var
  return $ zip sample_fred sample_jill

generateGamePlot :: Int -> Double -> IO()
generateGamePlot nr var = do
  sample <- sampleGames nr var
  let title = winRatio sample
  skillPlot title var sample


task4 :: IO()
task4 = do
  generateGamePlot 10000 5
  generateGamePlot 10000 2
  generateGamePlot 10000 10

posteriorWinPropablity :: Double -> Double -> Double -> Double
posteriorWinPropablity jskill fskill std = normcdf ( (jskill-fskill)/( sqrt 2 * std))

task5 :: IO ()
task5 = do
  print $ posteriorWinPropablity 15 12.5 5
  print $ posteriorWinPropablity 15 12.5 2
  print $ posteriorWinPropablity 15 12.5 10

normalPDF :: Double -> Double -> [Double] -> [(Double, Double)]
normalPDF mu std xs  =
  [ (x, 1/ (sqrt (2*pi* (std^2))) * exp (-(x-mu)^2/(2 * std^2 ) ) ) | x <- xs ]


plotAverageConvolution :: Int -> IO()
plotAverageConvolution nr_average = do
  mus <- sampleNormal nr_average 100 5
  let normals = map (\mu -> normalPDF mu 5 [0,0.1..140]) mus
  let nr = int2Double nr_average
  let average = map (\xs -> (fst $ head xs, sum ( map snd xs )/nr )) $ transpose normals
  let title = "../visualisation/average_gaussion_" ++ show nr_average   ++ ".png"
  toFile def title $ plot ( line "" [average])
  

task6 :: IO ()
task6 = do
  plotAverageConvolution 3
  plotAverageConvolution 6
  plotAverageConvolution 100

plotDists cumul pdf product = toFile def "../visualisation/pdf_to_approx.png" $ do
  plot( line "cumulative input" [cumul] )
  plot ( line "sensitive pdf" [pdf])
  plot ( line "product" [product])

norm_sample :: [(Double,Double)] -> [(Double,Double)]
norm_sample sample = map (\(a,b) -> (a,b/area)) sample
  where 
    area = sum $ map snd sample

task7 :: IO ()
task7 = do
  let skill_fred = ("fred", 100::Double, 5::Double)
  let dist_perf = ("perf", 0::Double, 5::Double)
  let conv1 = convolution dist_perf skill_fred 
  let skill_jill = ("jill", 120, 40)
  let (_ , mu, std) = convolution dist_perf skill_jill
  print mu
  print std
  let cumul = cumulative conv1 True
  let sample = norm_sample $ map(\x -> (x, cumul x)) [-50..300]
  let pdf = norm_sample $ normalPDF mu 40 [-50..300]
  let product = norm_sample $ zipWith (\(a,b) (c,d) -> (a,b*d)) sample pdf
  plotDists sample pdf product
  
task8 :: IO ()
task8 = do 
  let skill_fred = ("fred", 100::Double, 5::Double)
  let dist_perf = ("perf", 0::Double, 5::Double)
  let conv1 = convolution dist_perf skill_fred 
  let skill_jill = ("jill", 120, 40)
  let (_ , mu, std) = convolution dist_perf skill_jill
  let cumul = cumulative conv1 True
  let sample = norm_sample $ map(\x -> (x, cumul x)) [-50..300]
  let pdf = norm_sample $ normalPDF mu std [-50..300]
  let product = norm_sample $ zipWith (\(a,b) (c,d) -> (a,b*d)) sample pdf
  let mean = sum $ map (\(a,b) -> a*b) product
  let mean_of_square = sum $ map (\(a,b) -> a^2*b) product
  let std = sqrt $ mean_of_square - mean^2
  let (_,mux,stdx) = normalDivision ("test", mean, std) skill_jill
  print mux
  print stdx

plotDivide input = toFile def "../visualisation/divide.png" $ plot ( line "test" [input])

test :: IO ()
test = do
  let normal1 = normalPDF 120 5 [100..140]
  let normal2 = normalPDF 120 40 [100..140]
  let divide = map (\((a,b),(c,d)) -> if d == 0 then (a,0) else (a, b /d)) $ zip normal2 normal1
  plotDivide divide

nodes =
        [ GF "Fred_Skill" 100 5
        , GF "Jill_Skill" 120 40
        , DGF "Fred_perf" "Fred_Skill" 5
        , DGF "Jill_perf" "Jill_Skill" 5
        , GrT "Jill_Win" ("Fred_perf", "Jill_perf") 0.5
        ]

graph = generateGraph nodes 

plotPriorVsPost :: String -> [(Double,Double)] -> [(Double,Double)] -> [(Double,Double)] -> [(Double,Double)]-> IO ()
plotPriorVsPost title fredPrior fredPosterior jillPrior jillPosterior = toFile def ("../visualisation/" ++ title) $ do
  layout_title .= title
  plot (line "Fred Prios" [fredPrior])
  plot (line "Fred Posterior" [fredPosterior])
  plot (line "Jill Prios" [jillPrior])
  plot (line "Jill Posterior" [jillPosterior])

task9 :: IO ()
task9 = do
  let [(fredSkill,mu1,std1),(jillSkill,mu2,std2)] =  snd $ evalVariables ["Fred_Skill","Jill_Skill"] ["Jill_Win"] [[True]] graph
  let [(fredSkill',mu1',std1'),(jillSkill',mu2',std2')] =  snd $ evalVariables ["Fred_Skill","Jill_Skill"] ["Jill_Win"] [[False]] graph

  let fredPrior = normalPDF 100 5 [-30..300]
  let jillPrior = normalPDF 120 40 [-30..300]
  -- Jill wins
  let fredPosterior = normalPDF mu1 std1 [-30..300]
  let jillPosterior = normalPDF mu2 std2 [-30..300]
  --  Jill looses
  let fredPosterior' = normalPDF mu1' std1' [-30..300]
  let jillPosterior' = normalPDF mu2' std2' [-30..300]
  plotPriorVsPost "Priors_vs_Posterios_Jill_Win.png" fredPrior fredPosterior jillPrior jillPosterior
  plotPriorVsPost "Priors_vs_Posterios_Jill_Loose.png" fredPrior fredPosterior' jillPrior jillPosterior'

evalGameEntry :: Int -> IO ()
evalGameEntry nr = do
  (id1,id2,win_id, mu1, std1, mu2, std2) <- getGameRecord nr
  let upGraph = replaceVariables [("Fred_Skill",mu1,std1),("Jill_Skill",mu2,std2)] graph
  let [(fredSkill,mu1',std1'),(jillSkill,mu2',std2')] =  snd $ evalVariables ["Fred_Skill","Jill_Skill"] ["Jill_Win"] [[id2 == win_id]] upGraph
  updatePlayer id1 mu1' std1'
  updatePlayer id2 mu2' std2'
  --print $ show nr ++ " " 
  --print $ show id1 ++ " " ++ show mu1 ++ " " ++ show std1 
  --print $ show id1 ++ " " ++ show mu1' ++ " " ++ show std1'
  --print $ show id1 ++ " " ++ show mu2 ++ " " ++ show std2 
  --print $ show id2 ++ " " ++ show mu2' ++ " " ++ show std2'

  

testingTheFakeDataBase :: IO ()
testingTheFakeDataBase = do
  initDatabase
  nrGames <- numberGames
  mapM_ evalGameEntry [1..nrGames]
  print "Skill prediction after 10000 observed Game results:"
  mapM_ printPlayer [1..100]
  resetPlayers 


main :: IO ()
main = do
  task1
